### PR TITLE
1000x faster swap mapping and Cython build chain

### PR DESCRIPTION
**Pull request by nonhermitian**
### Summary
This PR refactors the `StochasticSwap` transpiler pass, and rewrites the core swap trial in Cython.  This results in a 1000x improvement in the swap trial computation (at least for Quantum Volume circuits).  In addition, this PR will put a Cython build into the `setup.py` file.

### Details and comments

For comparison, a 20x20 quantum volume circuit compiles in 16sec using this PR.  The standard compiler chain takes about 65 sec (~4x speedup).  However, running the profiler, gives the amount of time spent computing the actual swaps to be 0.13 vs 106 sec (slower since profiling), or ~800x speedup.

### CODE REVIEW

**Comment by nonhermitian** [comment]
@mtreinish This is now working, save for the tests mentioned in #1794, and the need to build the Cython files yourself at the moment.  I usually just have `setup` `cythonize` the `.pyx` files and then build them.  This would of course require having `cython` as a dependency.  See the `setup` in the created cython directory.  The other option would be to have a utility that generates the source, called by devs only, and the `setup` just builds the source like it would any other extension.

The cython files themselves are free of any `blas` or `cimport numpy` calls, so they can easily be pre-built and distributed in a wheel.

**Comment by nonhermitian** [comment]
Ok, the build process is working, and only the two known errors pop up.  I reverted back to using the NumPy RNG so that we have deterministic randoms.  The overhead is minimal.  However the two tests still fail due to the bug fix. I will verify the answers returned are good, and then change the expected output.

File: setup.py [hunk 1]
```diff
@@ -23,12 +28,49 @@
...
+
+INCLUDE_DIRS = []
+# Extra link args
+LINK_FLAGS = []
+# If on Win and Python version >= 3.5 and not in MSYS2 (i.e. Visual studio compile)
+if (sys.platform == 'win32' and os.environ.get('MSYSTEM') is None):
+    COMPILER_FLAGS = ['/w', '/O2', '/std:c++11']
+# Everything else
+else:
+    COMPILER_FLAGS = ['-w', '-O2', '-funroll-loops', '-std=c++11']
```
[the remark this answers is not in the document]
Comment (atilag) [reply setup.py]: `-ffast-math` could help here

[the remark this answers is not in the document]
Comment (nonhermitian) [reply setup.py]: I was using `-O3` previously.  Depending on who you talk to `-O2` is enough. e.g. even OpenBLAS compiles with `-O2` by default. The difference is not noticeable here.  I can turn on the warnings.  Also, `-ffast-math` is not going to do much here as the only real "math" is multiplication of doubles.

**Comment by nonhermitian** [comment]
@mtreinish this is back up and working.  Turned out that a remove underscore change was messing up the Cython build, and I needed to add cython and the build_ext call to AppVeyor.

**Comment by mtreinish** [comment]
I'm testing this locally to make sure we've got a seamless migration to include the cython bits. So far the first issue I've hit is on a clean virtualenv I pip install the repo with this patch applied and it fails on Cython not being found:

```
Processing /home/mtreinish/git/qiskit/qiskit-terra
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-r47nil_s/setup.py", line 14, in <module>
        from Cython.Build import cythonize
    ModuleNotFoundError: No module named 'Cython'
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-req-build-r47nil_s/
```

At a minimum we probably should add Cython to the `setup_requires` and `install_requires` in the setup() (https://setuptools.readthedocs.io/en/latest/setuptools.html ). But, I tried this locally and it doesn't ensure cython is installed before processing and hitting that error. We might just have to modify the install instructions to say ensure cython is installed first. I will push up my patch adding Cython to setup_requires just to clearly advertise that, even if it doesn't work as I hoped.

**Comment by nonhermitian** [comment]
Yes, you must have it installed beforehand.  There is a way to do that, but it is hacky.  Since we would presumably ship the binaries, I wasn'y too concerned about it since it would be a dev only requirement.

File: setup.py [hunk 2]
```diff
@@ -89,6 +89,7 @@
     keywords="qiskit sdk quantum",
     packages=find_packages(exclude=['test*']),
     install_requires=REQUIREMENTS,
+    setup_requires=['Cython>=0.29.5'],
```
Comment (nonhermitian) [inline setup.py]: Do we really need that high a version here?  I think `0.27.1` or higher should work.

Comment (mtreinish) [reply setup.py]: Sure, we can change it. I just picked the version I had installed locally when I was testing which I guess was the latest.

**Comment by mtreinish** [comment]
Sure I'm fine with that, we just need to make sure we update the docs to reflect that (and also the changelog). I'll push a patch updating the README, contributing guide, or wherever people have decided to hide the installing from source instructions to include this.

File: setup.py (hunk 2 above)
Comment (nonhermitian) [reply setup.py]: ok cool. And is it 'Cython' and not 'cython'?

Comment (mtreinish) [reply setup.py]: The package on pypi is `Cython`: https://pypi.org/project/Cython/ although pip is case insensitive so it doesn't really matter.

Comment (nonhermitian) [reply setup.py]: Ok, just double checking.  It is `cython` in conda.

**Pull request merged**